src/functional_test_runner/lib/mocha/decorate_mocha_ui.js

```diff
@@ -0,0 +1,134 @@
+  function wrapSuiteFunction(name, fn) {
+    return wrapFunction(fn, {
+      before() {
+        if (suiteCount > 0 && suiteLevel === 0) {
+          throw new Error(`
+            Test files must only define a single top-level suite. Please ensure that
+            all calls to \`describe()\` are within a single \`describe()\` call in this file.
+          `);
+        }
+
+        suiteCount += 1;
+        suiteLevel += 1;
+      },
+      after() {
+        suiteLevel -= 1;
+      }
+    });
+  }
+
+  /**
+   *  Wrap test functions to emit "testFailure" lifecycle hooks
+   *  when they fail and throw when they are called outside of
+   *  a describe
+   *
+   *  @param  {String} name
+   *  @param  {Function} fn
+   *  @return {Function}
+   */
+  function wrapTestFunction(name, fn) {
+    return wrapNonSuiteFunction(name, wrapRunnableArgsWithErrorHandler(fn, async (err, test) => {
+      await lifecycle.trigger('testFailure', err, test);
+    }));
+  }
+
+  /**
+   *  Wrap test hook functions to emit "testHookFailure" lifecycle
+   *  hooks when they fail and throw when they are called outside
+   *  of a describe
+   *
+   *  @param  {String} name
+   *  @param  {Function} fn
+   *  @return {Function}
+   */
+  function wrapTestHookFunction(name, fn) {
+    return wrapNonSuiteFunction(name, wrapRunnableArgsWithErrorHandler(fn, async (err, test) => {
+      await lifecycle.trigger('testHookFailure', err, test);
+    }));
+  }
+
+  /**
+   *  Wrap all non describe() mocha ui functions to ensure
+   *  that they are not called outside of a describe block
+   *
+   *  @param  {String} name
+   *  @param  {Function} fn
+   *  @return {Function}
+   */
+  function wrapNonSuiteFunction(name, fn) {
+    return wrapFunction(fn, {
+      before() {
+        if (suiteLevel === 0) {
+          throw new Error(`
+            All ${name}() calls in test files must be within a describe() call.
+          `);
+        }
+      }
+    });
+  }
+
+  /**
+   *  called for every assignment while defining the mocha ui
+   *  and can return an alternate value that will be used for that
+   *  assignment
+   *
+   *  @param  {String} property
+   *  @param  {Any} value
+   *  @return {Any} replacement function
+   */
+  function assignmentInterceptor(property, value) {
+    if (typeof value !== 'function') {
+      return value;
+    }
+
+    switch (property) {
+      case 'describe':
+      case 'xdescribe':
+      case 'context':
+      case 'xcontext':
+        return wrapSuiteFunction(property, value);
+
+      case 'it':
+      case 'xit':
+      case 'specify':
+      case 'xspecify':
+        return wrapTestFunction(property, value);
+
+      case 'before':
+      case 'beforeEach':
+      case 'after':
+      case 'afterEach':
+      case 'run':
+        return wrapTestHookFunction(property, value);
+
+      default:
+        return wrapNonSuiteFunction(property, value);
+    }
+  }
+
+  return createAssignmentProxy(context, assignmentInterceptor);
+}
```

Comment:
Okay, so the original thing that was wrapping our methods was doing it one way for describe functions and another way for all non-describe functions. The new implementation cares about more types of functions because we want to enforce single top level suite and no functions outside of describes. Nice work